Flux commands for analyzing input data

fsteeg opened the pull request:
See https://gitlab.com/oersi/oersi-etl/-/issues/238

fsteeg commented:
The first module is ready for functional review, assigning @TobiasNx. Deployed to playground for testing: [your sample from #294, using the new `fix-list-paths` command](https://test.metafacture.org/playground/?flux=//+This+part+can+be+adjusted+to+your+own+data+source.+The+example+here+is+marc21%0A%22https%3A//raw.githubusercontent.com/metafacture/metafacture-core/master/metafacture-runner/src/main/dist/examples/read/marc21/10.marc21%22%0A%7C+open-http%0A%7C+as-lines%0A%7C+decode-marc21%0A%0A//+This+part+can+be+reused+for+all+incoming+data+streams+to+list+all+fix+paths.%0A%7C+fix-list-paths%0A%7C+print%0A%3B%0A&active-editor=fix).

From the documentation annotation:

> Lists all paths found in a record. These paths can be used in a Fix to address fields. Options: count (output occurence frequency of each path, sorted by highest frequency first; default: true), index (output individual repeated subfields and array elements with index numbers instead of '*'; default: false)

(If we count it's always sorted by highest frequency, because if we count we'll automatically sort internally, which would make it hard to have a consistent implementation with separate control over the sorting. I think this way it's both conceptually simple and hopefully covers your use cases.)

TobiasNx commented:
Implementation looks good.
Only the option `count=` should behave differently. Only the template should change and count-triples should not be dropped. So that you have the list of values/paths but not the counted occurence. The path/values would be uniq and no duplicate should be given:

something like this:
```
// This part can be adjusted to your own data source. The example here is marc21
"https://raw.githubusercontent.com/metafacture/metafacture-core/master/metafacture-runner/src/main/dist/examples/read/marc21/10.marc21"
| open-http
| as-lines
| decode-marc21

// This part can be reused for all incoming data streams to list all fix paths.
| fix("nothing()",repeatedFieldsToEntities="true", entityMemberName="*")
| flatten
| stream-to-triples
| count-triples(countBy="predicate")
| template("${s}")
| print;

```

TobiasNx commented:
Also perhaps the template should be `| template("${o}\t|\t${s}")`  ( when count=false then  `| template("${s}")`)

TobiasNx commented:
Great paths are unique now. +1
Concerning the template I am not sure if I prefer `[occurece] | [path]` more. This may be a simple  preference. So I am okay with that. +1
 
Concerning the sorting I thought about it again.

```
| stream-to-triples
| count-triples(countBy="predicate")
| template("${s}")
| print;
```

This does not sort by occurence automatically. How does your function deviate from this? The alphabetical sorting helps to get an understanding of the source data and its schema. Therefore I would prefer if we also would support alphabetical sorting too.

fsteeg commented:
> Concerning the template I am not sure if I prefer `[occurece] | [path]` more.

I'm only trying to implement your requirement, I don't have my own preference here. Could you give me an example of the desired output for both commands?

blackwinter commented:
Just a side note: Wouldn't it make sense to make the template configurable?

TobiasNx commented:
> > Concerning the template I am not sure if I prefer `[occurece] | [path]` more.
> 
> I'm only trying to implement your requirement, I don't have my own preference here. Could you give me an example of the desired output for both commands?

[path] | [occurrence]
6890 .5.*	|	18

vs
[occurrence] | [path]
18	|	6890 .5.*

TobiasNx commented:
> > > Concerning the template I am not sure if I prefer `[occurece] | [path]` more.
> > 
> > 
> > I'm only trying to implement your requirement, I don't have my own preference here. Could you give me an example of the desired output for both commands?
> 
> [path] | [occurrence] 6890 .5.* | 18
> 
> vs [occurrence] | [path] 18 | 6890 .5.*

but keep it the way you implement it

fsteeg commented:
> The alphabetical sorting helps to get an understanding of the source data and its schema.

I've changed it to sort alphabetically if `count` is false:

[fix-list-paths(count="false")](https://test.metafacture.org/playground/?flux=//+This+part+can+be+adjusted+to+your+own+data+source.+The+example+here+is+marc21%0A%22https%3A//raw.githubusercontent.com/metafacture/metafacture-core/master/metafacture-runner/src/main/dist/examples/read/marc21/10.marc21%22%0A%7C+open-http%0A%7C+as-lines%0A%7C+decode-marc21%0A%0A//+This+part+can+be+reused+for+all+incoming+data+streams+to+list+all+fix+paths.%0A%7C+fix-list-paths%28count%3D%22false%22%29%0A%7C+print%0A%3B%0A&active-editor=fix)

> Concerning the template I am not sure if I prefer `[occurece] | [path]` more.

I've changed it to use `${o}\t|\t${s}` as a common default template for both commands:

[fix-list-paths](https://test.metafacture.org/playground/?flux=//+This+part+can+be+adjusted+to+your+own+data+source.+The+example+here+is+marc21%0A%22https%3A//raw.githubusercontent.com/metafacture/metafacture-core/master/metafacture-runner/src/main/dist/examples/read/marc21/10.marc21%22%0A%7C+open-http%0A%7C+as-lines%0A%7C+decode-marc21%0A%0A//+This+part+can+be+reused+for+all+incoming+data+streams+to+list+all+fix+paths.%0A%7C+fix-list-paths%0A%7C+print%0A%3B%0A&active-editor=fix)
[fix-list-values("68900.0.*")](https://test.metafacture.org/playground/?flux=//+This+part+can+be+adjusted+to+your+own+data+source.+The+example+here+is+marc21%0A%22https%3A//raw.githubusercontent.com/metafacture/metafacture-core/master/metafacture-runner/src/main/dist/examples/read/marc21/10.marc21%22%0A%7C+open-http%0A%7C+as-lines%0A%7C+decode-marc21%0A%0A//+This+part+can+be+reused+for+all+incoming+data+streams+to+list+all+fix+paths.%0A//%7C+fix-list-paths%0A%7C+fix-list-values%28%2268900.0.%2A%22%29%0A%7C+print%0A%3B%0A&active-editor=fix)

With an option to set it (thanks for the input @blackwinter):

[fix-list-paths(template="${s}\t${o}")](https://test.metafacture.org/playground/?flux=//+This+part+can+be+adjusted+to+your+own+data+source.+The+example+here+is+marc21%0A%22https%3A//raw.githubusercontent.com/metafacture/metafacture-core/master/metafacture-runner/src/main/dist/examples/read/marc21/10.marc21%22%0A%7C+open-http%0A%7C+as-lines%0A%7C+decode-marc21%0A%0A//+This+part+can+be+reused+for+all+incoming+data+streams+to+list+all+fix+paths.%0A%7C+fix-list-paths%28template%3D%22%24%7Bs%7D\t%24%7Bo%7D%22%29%0A%7C+print%0A%3B%0A&active-editor=fix)

TobiasNx commented:
Thanks, thats all great!

blackwinter commented on metafix/build.gradle:
```diff
@@ -36,6 +36,8 @@ dependencies {
   implementation "org.metafacture:metafacture-io:${versions.metafacture}"
   implementation "org.metafacture:metafacture-javaintegration:${versions.metafacture}"
   implementation "org.metafacture:metafacture-mangling:${versions.metafacture}"
+  implementation "org.metafacture:metafacture-formatting:${versions.metafacture}"
```
Minor: Alphabetical sorting would be nice.

blackwinter commented on metafix/src/main/java/org/metafacture/metafix/MetafixListPaths.java:
```diff
@@ -0,0 +1,52 @@
+ * @author Fabian Steeg
+ */
+@Description("Lists all paths found in the input records. These paths can be used in a Fix to address fields. Options: " +
+        "count (output occurence frequency of each path, sorted by highest frequency first; default: true), " +
+        "template (for formatting the internal triple structure; default: ${o}\t|\t${s} if count is true, else ${s})" +
+        "index (output individual repeated subfields and array elements with index numbers instead of '*'; default: false)")
```
Should we use Markdown formatting here? (See [previous](https://github.com/metafacture/metafacture-core/pull/468#discussion_r975402708) discussion.)

blackwinter commented on metafix/src/main/java/org/metafacture/metafix/MetafixListValues.java:
```diff
@@ -0,0 +1,49 @@
+ * @author Fabian Steeg
+ */
+@Description("Lists all values found for the given path. The paths can be found using fix-list-paths. Options: " +
+        "count (output occurence frequency of each value, sorted by highest frequency first; default: true)" +
+        "template (for formatting the internal triple structure; default: ${o}\t|\t${s} if count is true, else ${s})")
```
Ditto re: Markdown formatting.

blackwinter commented on metafix/src/main/java/org/metafacture/metafix/MetafixListPaths.java:
```diff
@@ -0,0 +1,52 @@
+        "index (output individual repeated subfields and array elements with index numbers instead of '*'; default: false)")
+@In(StreamReceiver.class)
+@Out(String.class)
+@FluxCommand("fix-list-paths")
```
I assume that these commands are supposed to be namespaced to the metafacture-fix project (hence the `fix-` prefix), but especially with regard to the (almost definitely) upcoming merge into metafacture-core: Shouldn't the command names reflect the usual "verb-object" order? So `list-fix-paths` ("list" being the verb, "Fix paths" being the object).

blackwinter commented on metafix/src/main/java/org/metafacture/metafix/MetafixListValues.java:
```diff
@@ -0,0 +1,49 @@
+        "template (for formatting the internal triple structure; default: ${o}\t|\t${s} if count is true, else ${s})")
+@In(StreamReceiver.class)
+@Out(String.class)
+@FluxCommand("fix-list-values")
```
Ditto re: command name (i.e. `list-fix-values`).

blackwinter commented on metafix/src/main/java/org/metafacture/metafix/MetafixListPaths.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    public void setIndex(final boolean index) {
+        super.getFix().setEntityMemberName(index ? Metafix.DEFAULT_ENTITY_MEMBER_NAME : "*");
```
`super` is unnecessary here, isn't it?

blackwinter commented on metafix/src/main/java/org/metafacture/metafix/MetafixListPaths.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    public boolean getIndex() {
+        return super.getFix().getEntityMemberName().equals(Metafix.DEFAULT_ENTITY_MEMBER_NAME);
```
Ditto.

blackwinter commented on metafix/src/main/java/org/metafacture/metafix/MetafixStreamAnalyzer.java:
```diff
@@ -0,0 +1,142 @@
+    private static final String DEFAULT_COUNTED_TEMPLATE = "${o}\t|\t${s}";
+    private static final String DEFAULT_UNCOUNTED_TEMPLATE = "${s}";
+
+    private Metafix fix;
+    private boolean count = true;
+    private Compare countBy;
```
`fix` and `countBy` can be made `final`, right?

blackwinter commented on metafix/src/main/java/org/metafacture/metafix/MetafixStreamAnalyzer.java:
```diff
@@ -0,0 +1,142 @@
+        return this.fix;
+    }
+
+    /* package-private */ void setFix(final Metafix fix) {
```
This method is unused, right?

blackwinter commented on metafix/src/test/java/org/metafacture/metafix/MetafixListValuesTest.java:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    private void verify(final String... result) throws MockitoAssertionError {
+        lister.setTemplate(lister.getCount() ? "${o}\t|\t${s}" : "${s}");
```
Is this line necessary? We are/should be testing the default template(s).

blackwinter reviewed (changes requested):
Left some (more or less) minor remarks.

fsteeg commented:
Thank you for the review! Addressed your feedback in [172a701](https://github.com/metafacture/metafacture-fix/pull/300/commits/172a701addefcbdc4cfece9d687c2469e60f292a).
